grafana/grafana
# loki: backend-mode: add nanosecond-timestamps

loki returns logs-data with timestamps that are nanosecond-precise. the normal timestamps we use in grafana are only millisecond-precise, so for logs, we simply return an additional dataframe-field, where there are timestamps as nanosecond-unix-timestamps-as-strings.

the way this currently works in grafana, is that this field is marked `FieldType.time`, so we have to do this adjustment in javascript. (NOTE: i am looking into an alternative way to do this, but this is how the current-frontend-code does it so this is how we do it too for now)

(i also corrected the description-text in an unrelated test to avoid confusion)

how to test:
1. make sure you have the `lokiBackendMode` feature-flag enabled
2. go to explore-mode, run a logs-producing loki query
3. verify in the browser-developer-tools that the ajax request is going to `/api/ds/query`, not to `/api/datasources/proxy/...`
4. open the browser devtools, go to the "Sources" tab, find the file `logs_model.ts` (on a mac pressing cmd+p opens the file browser), and place a breakpoint here: https://github.com/grafana/grafana/blob/main/public/app/core/logs_model.ts#L372
5. run the loki query again, you will be paused on that line. mouse-hover over the `tsNs` variable, and verify that it is not `undefined`
